### PR TITLE
Add ghe config option

### LINKED ISSUE #552: Interested in support for self-hosted GHE installation
Hi Allstar Team,

We are currently evaluating the use of OSSF Allstar on our self-hosted Mercedes-Benz GitHub Enterprise instance.

Before we could get it to work, we had to modify all the locations where GitHub clients are created to configure them to use our internal URL, as there is currently no option to set this via configuration or parameter.

We believe that there may be more self-hosted GitHub Enterprise cases beyond our company, and we would like to know if you would be interested in enhancing OSSF Allstar to support this scenario.

If this is of interest to you, we would be happy to get in touch and contribute to the development of this feature.

### PR DESCRIPTION
**Pull request by SebastianBezold**
## Description

This PR introduces support for self hosted GitHub Enterprise instances with URLs different than github.com.
With these changes, an environment variable `ALLSTAR_GHE_URL` can used to configure a custom GitHub URL.
If set, all GitHub clients created are initialized using this URL.

## Additional Info

For some reason, this setup only works if additionally to `ALLSTAR_GHE_URL`, you also define `GH_HOST` with the same host, just except the protocol (i.e: `ALLSTAR_GHE_URL="https://my-ghe.example.com"` and `GH_HOST="my-ghe.example.com"`).

I have tested this PR on our own internal GitHub enterprise instance with `KEY_SECRET="direct"` and providing the `PRIVATe_KEY` as plain-text `.pem`

Closes #552 

<sub>Sebastian Bezold sebastian.bezold@mercedes-benz.com, Mercedes-Benz Tech Innovation GmbH, [legal info/impressum](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)</sub>

### CODE REVIEW

File: operator.md [hunk 1]
```diff
@@ -55,12 +55,27 @@ conditions on enforcement actions, ex: pinging an issue twice at the same time.
...
+In case you want to operate Allstar with a self-hosted GitHub Enterprise instance, you need to set the `ALLSTAR_GHE_URL` environment variable to the URL of your GitHub Enterprise instance URL.
+The different API endpoints for API and upload are appended automatically.
+
+Example: 
+
+Given, your GHE instance URL is "https://my-ghe.example.com", you need to set the following environment variables:
+
+```shell
+export ALLSTAR_GHE_URL="https://my-ghe.example.com"
+export GH_HOST="my-ghe.example.com"  # This is somehow used within a dependency and (as of now) is unclear how to set it programmatically.
```
Comment (spencerschrock) [inline operator.md]: `GH_HOST` is used by the Scorecard dependency when setting up our clients to interact with GitHub. We don't currently have a way of setting this without using an environment variable. If this is a problem, I'm sure we could consider alternatives.

https://github.com/ossf/scorecard/blob/4303b741ea5afbb9b0d0420788043defb24f3740/clients/githubrepo/repo.go#L123-L132

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@ cmd/allstar/allstar
 *.pem
 .vscode
 allstar.ref
+.idea/
```
Comment (jeffmendoza) [inline .gitignore]: Add newline at end of file

File: pkg/ghclients/ghclients.go [hunk 3]
```diff
@@ -92,19 +93,45 @@ func (g *GHClients) Get(i int64) (*github.Client, error) {
 	}
 
 	var tr http.RoundTripper
-	var err error
 	if i == 0 {
-		tr, err = ghinstallationNewAppsTransport(ctr, operator.AppID, g.key)
+		appTransport, _ := ghinstallationNewAppsTransport(ctr, operator.AppID, g.key)
```
Comment (jeffmendoza) [inline pkg/ghclients/ghclients.go]: Check error

File: pkg/ghclients/ghclients.go [hunk 4]
```diff
@@ -92,19 +93,45 @@ func (g *GHClients) Get(i int64) (*github.Client, error) {
...
+		// other than clien.WithEnterpriseUrls, setting the BaseUrl plainly, we need to ensure the /api/v3 ending
+		appTransport.BaseURL = fullEnterpriseApiUrl(operator.GitHubEnterpriseUrl)
```
Comment (jeffmendoza) [inline pkg/ghclients/ghclients.go]: Wrap in `if operator.GitHubEnterpriseUrl != "" {`

File: pkg/ghclients/ghclients.go [hunk 5]
```diff
@@ -92,19 +93,45 @@ func (g *GHClients) Get(i int64) (*github.Client, error) {
...
+		tr = appTransport
 	} else {
-		tr, err = ghinstallationNew(ctr, operator.AppID, i, g.key)
+		ghiTransport, _ := ghinstallationNew(ctr, operator.AppID, i, g.key)
```
Comment (jeffmendoza) [inline pkg/ghclients/ghclients.go]: Check error

**Review by jeffmendoza — COMMENTED** [review]
Thanks, looks great! Just a couple nits.

**Comment by SebastianBezold** [comment]
Hi @spencerschrock and @jeffmendoza,

sorry for the very late response. Have been on a longer private absence and was not able to work on this.
Thank you for your input. I'm right back working on it!

File: operator.md (hunk 1 above)
Comment (SebastianBezold) [reply operator.md]: IMHO, it is fine to just mention it in the docs. Thanks for the explanation 👍

**Comment by SebastianBezold** [comment]
Hi @jeffmendoza,

are the fixes to your comments ok? Is there anything else, that I should adapt?

**Review by jeffmendoza — APPROVED** [review]

**Comment by jeffmendoza** [comment]
Thanks for the reminder. Looks great. Thank you!!

**Pull request merged**